### PR TITLE
Support building and testing for noble

### PR DESCRIPTION
**Pull request by addyess**
### Overview

Adjusts the CI to begin testing against noble (24.04) as well as previously supported bionic, focal, and jammy

### Details
* continue to build a monolithic charm that supports all arches and series using charmcraft 2.x
* move metadata and config into charmcraft.yaml
* ensure code operates from 3.6 (bionic) -> 3.12 (noble)

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -43,17 +43,18 @@ jobs:
...
         with:
           python-version: '3.10'
-      - name: Install Dependencies
-        run: |
-          pip install tox
       - name: Setup operator environment
         uses: charmed-kubernetes/actions-operator@main
         with:
-          juju-channel: 3.1/stable
+          juju-channel: 3/stable
```
Comment (aieri) [inline .github/workflows/tests.yml:57]: juju 3 does not support bionic so that's probably why the tests fail (no idea why they didn't previously though)

**Review by aieri — COMMENTED** [review]
if we want to continue supporting 18.04 we'd need to test with juju 2.9 as well. How about dropping bionic? The older revision will still live in the store

File: .github/workflows/tests.yml (hunk 1 above)
Comment (addyess) [reply .github/workflows/tests.yml:57]: it looks like bionic didn't come alive b/c of the version of python didn't mesh well with the newer version of ops.  I'm backing up to `ops<2.0` to see how well that goes.

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -3,27 +3,28 @@
...
 
 
 log = logging.getLogger(__name__)
-meta = yaml.safe_load(Path("metadata.yaml").read_text())
+meta = yaml.safe_load(Path("charmcraft.yaml").read_text())
+CHANNEL_TO_SERIES = dict(reversed(mapping) for mapping in ALL_SERIES_VERSIONS.items())
+CHARM_SUPPORT = {
+    CHANNEL_TO_SERIES[base["channel"]]: base["channel"]
+    for base in meta["bases"][0]["run-on"]
+}
```
Comment (addyess) [inline tests/integration/test_charm.py:15]: reads the "runs-on" from the first base and tests for each `series`

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -1,10 +1,22 @@
...
 
+def charm_config() -> str:
+    """Return the charm configuration as a string readfrom charmcraft.yaml."""
+    filename = inspect.getfile(charm.UbuntuCharm)
+    charm_dir = Path(filename).parents[1]
+    content = (charm_dir / "charmcraft.yaml").read_text()
+    config = yaml.safe_load(content)["config"]
+    return yaml.safe_dump(config)
+
+
```
Comment (addyess) [inline tests/unit/test_charm.py:19]: `ops>2` doesn't require this -- but in order to maintain bionic, we can't use `ops>2`

File: tox.ini [hunk 4]
```diff
@@ -14,7 +14,7 @@ setenv =
 deps =
     pytest
     ipdb
-    ops
+    -r{toxinidir}/requirements.txt
```
Comment (addyess) [inline tox.ini:17]: Make sure the unit tests use the charm requirements.txt

**Comment by addyess** [comment]
@javacruft Could you review this magnificent upgrade to the ubuntu charm

**Pull request merged**